### PR TITLE
Improved backuping

### PR DESCRIPTION
**Pull request by atengler**
Improved automated backuping of PostgreSQL databases

### CODE REVIEW

**Review by fpytloun — CHANGES_REQUESTED** [review]
Totally wrong, hard-coded stuff, abusing home directory for temp files, etc.

**Comment by michaelkuty** [comment]
Probably you mean restoring :D

File: postgresql/files/restore.sh [hunk 1]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+export AWS_ACCESS_KEY_ID={{ database.initial_data.auth.awsaccesskeyid }}
+export AWS_SECRET_ACCESS_KEY={{ database.initial_data.auth.awssecretaccesskey }}
+{%- endif %}
 
+{%- if database.initial_data.engine == "backupninja" %}
+scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
 gunzip -d -1 -f {{ target }}{{ dest_name }}
-
 pg_restore /root/postgresql/data/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
+{%- else %}
```
Comment (epcim) [inline postgresql/files/restore.sh:45]: else is not appropriate, use `if database.initial_data.engine == "s3"`

File: postgresql/files/restore.sh [hunk 2]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
 {%- set dest_name = db_name + ".pg_dump.gz" %}
 {%- set target = "/root/postgresql/data/" %}
```
Comment (epcim) [inline postgresql/files/restore.sh:34]: avoid to use /root as temporary folder, use /tmp

File: postgresql/files/restore.sh [hunk 3]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+{%- if database.initial_data.auth is defined %}
+export AWS_ACCESS_KEY_ID={{ database.initial_data.auth.awsaccesskeyid }}
+export AWS_SECRET_ACCESS_KEY={{ database.initial_data.auth.awssecretaccesskey }}
+{%- endif %}
 
+{%- if database.initial_data.engine == "backupninja" %}
+scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
 gunzip -d -1 -f {{ target }}{{ dest_name }}
-
 pg_restore /root/postgresql/data/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
```
Comment (epcim) [inline postgresql/files/restore.sh:44]: Fix hardcoded path. Use temp directory for this + cleanup.

File: postgresql/files/restore.sh [hunk 4]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+{%- else %}
+duplicity --s3-unencrypted-connection --s3-use-new-style --s3-european-buckets s3+http://{{ database.initial_data.source }}/{{ host }} {{ target }}{{ dest_name }} 
+gunzip -d -1 -f {{ target }}{{ dest_name }}/var/backups/postgresql/{{ source_name }}
+pg_restore /root/postgresql/data/{{ dest_name }}/var/backups/postgresql/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
+{%- endif %}
 
 touch /root/postgresql/flags/{{ db_name }}-installed
```
Comment (epcim) [inline postgresql/files/restore.sh:51]: This is pretty poor way.
- better to use "postgresql" user directory.
- better to include timestamp

Before file is created, validate db with:
https://serverfault.com/questions/175865/how-to-check-postgresql-database-integrity

File: postgresql/files/restore.sh [hunk 5]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+{%- endif %}
 
+{%- if database.initial_data.engine == "backupninja" %}
+scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
 gunzip -d -1 -f {{ target }}{{ dest_name }}
-
 pg_restore /root/postgresql/data/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
+{%- else %}
+duplicity --s3-unencrypted-connection --s3-use-new-style --s3-european-buckets s3+http://{{ database.initial_data.source }}/{{ host }} {{ target }}{{ dest_name }} 
+gunzip -d -1 -f {{ target }}{{ dest_name }}/var/backups/postgresql/{{ source_name }}
```
Comment (epcim) [inline postgresql/files/restore.sh:47]: exit if this gunzip fails

File: postgresql/files/restore.sh [hunk 6]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+pg_restore /root/postgresql/data/{{ dest_name }}/var/backups/postgresql/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
```
Comment (epcim) [inline postgresql/files/restore.sh:48]: allow restore to alternative db name?

Comment (epcim) [inline postgresql/files/restore.sh:48]: hardcoded path

File: postgresql/files/restore.sh [hunk 7]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
+export AWS_SECRET_ACCESS_KEY={{ database.initial_data.auth.awssecretaccesskey }}
+{%- endif %}
 
+{%- if database.initial_data.engine == "backupninja" %}
+scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
 gunzip -d -1 -f {{ target }}{{ dest_name }}
-
 pg_restore /root/postgresql/data/{{ db_name }}.pg_dump --dbname={{ db_name }} --host=localhost --username={{ database.users[0].name }} --no-password -c
+{%- else %}
+duplicity --s3-unencrypted-connection --s3-use-new-style --s3-european-buckets s3+http://{{ database.initial_data.source }}/{{ host }} {{ target }}{{ dest_name }}
```
Comment (epcim) [inline postgresql/files/restore.sh:46]: make s3+http part of the initial_data.source as it may be "s3+https"

File: postgresql/files/restore.sh [hunk 8]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
...
 {%- set target = "/root/postgresql/data/" %}
 
-scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
+{%- if database.initial_data.auth is defined %}
+export AWS_ACCESS_KEY_ID={{ database.initial_data.auth.awsaccesskeyid }}
+export AWS_SECRET_ACCESS_KEY={{ database.initial_data.auth.awssecretaccesskey }}
+{%- endif %}
 
+{%- if database.initial_data.engine == "backupninja" %}
+scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }}
```
Comment (epcim) [inline postgresql/files/restore.sh:42]: hardcoded path

File: postgresql/files/restore.sh [hunk 9]
```diff
@@ -33,11 +33,20 @@ touch /root/postgresql/flags/{{ database.name }}-installed
 {%- set dest_name = db_name + ".pg_dump.gz" %}
 {%- set target = "/root/postgresql/data/" %}
 
-scp backupninja@{{ database.initial_data.source }}:/srv/backupninja/{{ host }}/var/backups/postgresql/postgresql.{{ age }}/{{ source_name }} {{ target }}{{ dest_name }} 
+{%- if database.initial_data.auth is defined %}
+export AWS_ACCESS_KEY_ID={{ database.initial_data.auth.awsaccesskeyid }}
```
Comment (epcim) [inline postgresql/files/restore.sh:37]: make s3 on aws conditional

**Review by epcim — CHANGES_REQUESTED** [review]
as of now do not provide level required

**Comment by fpytloun** [comment]
We shouldn't keep crappy pull requests opened to avoid @epcim reviewing them in a free time :-)
This script needs to be written from scratch, no need to kick into already dead horse.

**Pull request closed without merging**